caikit_nlp/modules/text_generation/fine_tuning.py

```diff
@@ -0,0 +1,262 @@
+)
+class FineTuning(ModuleBase):
+    """Module to provide fine-tuning support for text generation task"""
+
+    def __init__(self, tokenizer, model):
+        super().__init__()
+
+        self.tokenizer = tokenizer
+        # NOTE: self.model here can also be HF trainer. This is because
+        # if we have just trained the model then the models weights might be
+        # available in different devices (and configuration), depending on
+        # how it was trained. For now (July 10, 2023), we are not trying to
+        # extract the model out from trainer itself, since that would require
+        # us to essentially save it or reconstruct it to do normal inferring.
+        self.model = model
+
+    @classmethod
+    def train(
+        cls,
+        base_model: str,  # TODO: Union[str, PretrainedModelBase]
+        train_stream: DataStream[GenerationTrainRecord],
+        torch_dtype: str = None,  # TODO: Optional[Union[torch.dtype, str]]
+        max_source_length: int = 256,
+        max_target_length: int = 128,
+        batch_size: int = 8,
+        num_epochs: int = 5,
+        accumulate_steps: int = 32,
+        shuffle: bool = True,
+        lr: float = 2e-5,
+        checkpoint_dir: str = "/tmp",  # Directory where model predictions and checkpoints will be written
+    ):
+        """
+        # FIXME: Below is currently configured for Seq2Seq only
+        """
+
+        torch_dtype = get_torch_dtype(torch_dtype)
+
+        ## Load base model
+        if isinstance(base_model, str):
+            model_config = AutoConfig.from_pretrained(base_model)
+
+            resource_type = None
+            for resource in cls.supported_resources:
+                if model_config.model_type in resource.SUPPORTED_MODEL_TYPES:
+                    resource_type = resource
+                    break
+
+            if not resource_type:
+                error(
+                    "<NLP61784225E>",
+                    "{} model type is not supported currently!".format(
+                        model_config.model_type
+                    ),
+                )
+            log.debug("Bootstrapping base resource [%s]", base_model)
+            base_model = resource_type.bootstrap(base_model, torch_dtype=torch_dtype)
+
+        ## Generate data loader from stream
+        training_dataset: IterableDataset = cls._preprocess_function(
+            train_stream=train_stream,
+            tokenizer=base_model.tokenizer,
+            max_source_length=max_source_length,
+            max_target_length=max_target_length,
+            shuffle=shuffle,
+        )
+
+        ## TODO: Fetch trainer from resource
+
+        # TODO: Make this whole thing configurable by end-users, by optionally accepting `training_args`
+        # as argument to this train function.
+        # TODO: Remove all the default used below and make them all configurable
+        training_args = Seq2SeqTrainingArguments(
```

Comment:
I think I might misunderstand this - is this module not providing fine tuning support to causal LM also? Also, can we fallback to defaults in the trainer or something instead of hard coding default overrides here?

Reply:
this will eventually support trainings for various types of models, for this iteration, I just started with seq2seq. In next iteration, I'll move this whole logic to  individual `resources`, so we just do `.get_trainer()` on it.

I left a comment related to that [here](https://github.com/caikit/caikit-nlp/pull/72/files#diff-5e944359dd88fad4652c05a882bc242a6b567ded247a67ac24c9fc71ef9c263bR118)